internal/controller/object/object.go

```diff
@@ -340,3 +354,160 @@ func setObserved(obj *v1alpha1.Object, observed *unstructured.Unstructured) erro
+		}
+
+		// Patch fields if any
+		if err := ref.ApplyFromFieldPathPatch(res, obj); err != nil {
+			return errors.Wrap(err, errPatchFromReferencedResource)
+		}
+
+		// Add finalizer if not exists
+		finalizer := "object/" + string(obj.UID)
+		if !meta.FinalizerExists(res, finalizer) {
+			meta.AddFinalizer(res, finalizer)
+			if err := kubeClient.Update(ctx, res); err != nil {
+				return errors.Wrap(err, errFailedToAddFinalizer)
+			}
+		}
+	}
+
+	return nil
+}
+
+// removeReferenceFinalizers removes finalizers previously added to references.
+func (c *external) removeReferenceFinalizers(ctx context.Context, obj *v1alpha1.Object) {
+	c.logger.Debug("Removing finalizers from referencies.")
+
+	// Loop through references to resolve each referenced resource
+	for _, ref := range obj.Spec.References {
+		var kubeClient client.Client
+
+		if ref.FromObject.APIVersion == obj.APIVersion && ref.FromObject.Kind == obj.Kind {
+			// Object resource handled by provider on local cluster
+			kubeClient = c.localClient
+		} else {
+			// Other resource usually handled on remote cluster
+			// But could also be handled on local cluster
+			kubeClient = c.client
+		}
+
+		res := &unstructured.Unstructured{}
+		res.SetAPIVersion(ref.FromObject.APIVersion)
+		res.SetKind(ref.FromObject.Kind)
+		// Try to get referenced resource
+		err := kubeClient.Get(ctx, client.ObjectKey{
+			Namespace: ref.FromObject.Namespace,
+			Name:      ref.FromObject.Name,
+		}, res)
+
+		if err != nil {
+			c.logger.Debug(errGetReferencedResource, "error", err)
+			continue
+		}
+
+		// Remove finalizer if exists
+		finalizer := "object/" + string(obj.UID)
+		if meta.FinalizerExists(res, finalizer) {
+			meta.RemoveFinalizer(res, finalizer)
+			if err := kubeClient.Update(ctx, res); err != nil {
+				c.logger.Debug(errFailedToRemoveFinalizer, "error", err)
+				continue
+			}
+		}
+	}
+}
+
+func (c *external) isNotFound(ctx context.Context, obj *v1alpha1.Object, err error) bool {
+	isNotFound := false
+
+	if kerrors.IsNotFound(err) {
+		isNotFound = true
+	} else if meta.WasDeleted(obj) {
+		// If the Object resource was being deleted but the external resource is
+		// not deletable as management policy is specified, we should return the
+		// external resource not found, so that Object can be deleted by managed
+		// resource reconciler. Otherwise, the reconciler will try to delete the
+		// external resource which breaks the management policy.
+		if policy := obj.Spec.ManagementPolicy; policy == v1alpha1.ObserveCreateUpdate || policy == v1alpha1.Observe {
+			c.logger.Debug("Managed resource was deleted but external resource is undeletable.")
+			isNotFound = true
+		}
+	}
+
+	if meta.WasDeleted(obj) && isNotFound {
+		// If the external resource is not found, we should remove the finalizer
+		// that was previously added to the references of this Object, since the
+		// Object, as the dependant resource, was deleted.
+		c.removeReferenceFinalizers(ctx, obj)
```

Comment:
Better to move this to `Delete()` which would simplify this method with early returns.

Reply:
Code has been moved to `RemoveFinalizer()`.